adoptedStyleSheets.js

```diff
@@ -1,180 +1,324 @@
-      }
-      if (this[node]) {
-        this[node]._sheet.innerHTML = contents;
-        updateAdopters(this);
-        return this[node]._sheet.sheet;
-      } else {
-        throw new TypeError('replaceSync can only be called on a constructed style sheet');
-      }
+    for (const methodKey of cssStyleSheetMethods) {
+      // Here we apply all changes we have done to the original CSSStyleSheet
+      // object to all adopted style element.
+      const oldMethod = proto[methodKey];
+      proto[methodKey] = function(...args) {
+        if ($constructStyleSheet in this) {
+          for (const [, styleElement] of this[$constructStyleSheet].adopters) {
+            if (styleElement.sheet) {
+              styleElement.sheet[methodKey](...args);
+            }
+          }
+
+          // And we also need to remember all these changes to apply them to
+          // each newly adopted style element.
+          this[$constructStyleSheet].actions.push([methodKey, args]);
+        }
+
+        return oldMethod.apply(this, args);
+      };
+    }
+  };
+
+  const updateAdopters = sheet => {
+    for (const [, styleElement] of sheet[$constructStyleSheet].adopters) {
+      styleElement.innerHTML =
+        sheet[$constructStyleSheet].basicStyleElement.innerHTML;
+    }
+  };
+
+  const importPattern = /\@import/;
+
+  // This class will be a substitute for the CSSStyleSheet class that
+  // cannot be instantiated. The `new` operation will return the native
+  // CSSStyleSheet object extracted from a style element appended to the
+  // iframe.
+  class ConstructStyleSheet {
+    // Allows instanceof checks with the window.CSSStyleSheet.
+    static [Symbol.hasInstance](instance) {
+      return instance instanceof OldCSSStyleSheet;
     }
 
-    function replace(contents) {
+    constructor() {
+      // A style element to extract the native CSSStyleSheet object.
+      const basicStyleElement = document.createElement('style');
+      frameBody.append(basicStyleElement);
+
+      const nativeStyleSheet = basicStyleElement.sheet;
+      updatePrototype(nativeStyleSheet.constructor.prototype);
+
+      // A support object to preserve all the polyfill data
+      nativeStyleSheet[$constructStyleSheet] = {
+        adopters: new Map(),
+        actions: [],
+        basicStyleElement,
+      };
+
+      return nativeStyleSheet;
+    }
+
+    replace(contents) {
       return new Promise((resolve, reject) => {
-        if (this[node]) {
-          this[node]._sheet.innerHTML = contents;
-          resolve(this[node]._sheet.sheet);
+        if (this[$constructStyleSheet]) {
+          this[$constructStyleSheet].basicStyleElement.innerHTML = contents;
+          resolve(this[$constructStyleSheet].basicStyleElement.sheet);
           updateAdopters(this);
         } else {
-          reject('replace can only be called on a constructed style sheet');
+          reject(
+            new TypeError(
```

Comment:
Hmm, I'm not sure I follow. Should I just replace the
```javascript
reject(
  new TypeError(
    "Illegal invocation: 'replace' can only be called on a Constructible Style Sheet",
  ),
);
```
with:
```javascript
reject(
  new DOMException(
    "Failed to execute 'replace' on 'CSSStyleSheet': Can't call replace on non-constructed CSSStyleSheets.'",
    'NotAllowedError',
  ),
);
```
?

If I got it correctly, I'm not sure we should change the error type, because Chrome throws the `TypeError`.

Reply:
Oh, so the error is for this case. I misunderstood it purpose. Ok, I'll fix it.

Reply:
Fixed